numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -3694,6 +3693,7 @@ _vec_string_with_args(PyArrayObject* char_array, PyArray_Descr* type,
         if (item_result == NULL) {
             goto err;
         }
+        Py_DECREF(args_tuple);
```

Comment:
Ah, had not looked at the `err` tag! Still, you are decreasing the reference count of `args_tuple`, not of `item_result`, so I'm pretty sure you can `Py_DECREF` before the if without fear: `args_tuple` has been checked to not be `NULL` before the function call, and should remain so after it, regardless of whether it fails or suceeds.

I think it's nicer to handle the variable inside the loop where it is needed, i.e. move this up and remove the `Py_XDECREF` call, but if you have strong objections to that, I'm open to discussion.